File: java_ver/src/main/java/ru/HuffmanCompression.java
package ru;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

class HuffmanCompression {
    private String inputFileVersion = "1.0.0.0";
    private final List<TlkEntry> inputData = new LinkedList<>();
    private final Map<Character, Integer> frequencyCount = new HashMap<>();
    private final List<HuffmanNode> huffmanTree = new LinkedList<>();
    private final Map<Character, BitArray> huffmanCodes = new HashMap<>();

    private static class TlkEntry implements Comparable {
        public int stringID;
        public int position;
        public String data;

        public TlkEntry(int stringID, int position, String data) {
            this.stringID = stringID;
            this.position = position;
            this.data = data;
        }

        @Override
        public int compareTo(Object o) {
            return Integer.compare(this.position, ((TlkEntry) o).position);
        }

    }

    private static class HuffmanNode {
        public char data;
        public int frequencyCount;
        public HuffmanNode left;
        public HuffmanNode right;
        public int id;

        public HuffmanNode(char d, int freq) {
            data = d;
            frequencyCount = freq;
        }

        public HuffmanNode(HuffmanNode left, HuffmanNode right) {
            frequencyCount = left.frequencyCount + right.frequencyCount;
            this.left = left;
            this.right = right;
        }
    }

    /** <summary>
     *      Loads a file into memory and prepares for compressing it to TLK
     * </summary>
     * <param name="fileName"></param>
     * <param name="ff"></param>
     * <param name="debugVersion"></param>
     */
    public void loadInputData(String fileName, FileFormat ff, boolean debugVersion) throws IOException, ParserConfigurationException, SAXException {
        inputData.clear();
        LoadXmlInputData(fileName, debugVersion);
        Collections.sort(inputData);
        PrepareHuffmanCoding();
    }

    /** <summary>
     *       Dumps data from memory to TLK compressed file format.
     *       <remarks>
     *           Compressed data should be read into memory first, by LoadInputData method.
     *       </remarks>
     *  </summary>
     *  <param name="fileName"></param>
     *  <param name="isPC"></param>
     */
    public void saveToTlkFile(String fileName, boolean isPC) throws IOException {
        Files.deleteIfExists(Paths.get(fileName));

        /* converts Huffmann Tree to binary form */
        List<Integer> treeBuffer = ConvertHuffmanTreeToBuffer();

        /* preparing data and entries for writing to file
         * entries list consists of pairs <String ID, Offset> */
        List<BitArray> binaryData = new LinkedList<>();
        Map<Integer, Integer> entries1 = new HashMap<>();
        Map<Integer, Integer> entries2 = new HashMap<>();
        int offset = 0;

        for (TlkEntry entry : inputData) {
            if (entry.stringID < 0) {
                if (!entries1.containsKey(entry.stringID)) {
                    entries1.put(entry.stringID, Integer.parseInt(entry.data));
                } else {
                    entries2.put(entry.stringID, Integer.parseInt(entry.data));
                }
                continue;
            }

            if (!entries1.containsKey(entry.stringID)) {
                entries1.put(entry.stringID, offset);
            } else {
                entries2.put(entry.stringID, offset);
            }

            /* for every character in a string, put it's binary code into data array */
            for (char c : entry.data.toCharArray()) {
                binaryData.add(huffmanCodes.get(c));
                offset += huffmanCodes.get(c).length();
            }
        }

        /* preparing TLK Header */
        int magic = 7040084;
        int ver = 3;
        int min_ver = 2;
        int entry1Count = entries1.size();
        int entry2Count = entries2.size();
        int treeNodeCount = treeBuffer.size() / 2;
        int dataLength = offset / 8;
        if (offset % 8 > 0) {
            ++dataLength;
        }

        /* using LittleEndian for PC architecture and BigEndian for Xbox360 */
//        MiscUtil.Conversion.EndianBitConverter bitConverter;
//        if (isPC)
//            bitConverter = new MiscUtil.Conversion.LittleEndianBitConverter();
//        else
//            bitConverter = new MiscUtil.Conversion.BigEndianBitConverter();

        OutputStream bw = new FileOutputStream(fileName);

//        var bw = new MiscUtil.IO.EndianBinaryWriter(bitConverter, File.OpenWrite(fileName));

        /* writing TLK Header */
        bw.write(magic);
        bw.write(ver);
        bw.write(min_ver);
        bw.write(entry1Count);
        bw.write(entry2Count);
        bw.write(treeNodeCount);
        bw.write(dataLength);

        /* writing entries */
        for (Map.Entry<Integer, Integer> entry : entries1.entrySet()) {
            bw.write(entry.getKey());
            bw.write(entry.getValue());
        }
        for (Map.Entry<Integer, Integer> entry : entries2.entrySet()) {
            bw.write(entry.getKey());
            bw.write(entry.getValue());
        }

        /* writing HuffmanTree */
        for (int element : treeBuffer) {
            bw.write(element);
        }
        /* writing data */
        byte[] data = BitArrayListToByteArray(binaryData, offset);
        bw.write(data);
        bw.close();
    }

    /** <summary>
     *        Loads data from XML file into memory
     *  </summary>
     *  <param name="fileName"></param>
     *  <param name="debugVersion"></param>
     */
    private void LoadXmlInputData(String fileName, boolean debugVersion) throws IOException, SAXException, ParserConfigurationException {
        // Получение фабрики, чтобы после получить билдер документов.
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        // Получили из фабрики билдер, который парсит XML, создает структуру Document в виде иерархического дерева.
        DocumentBuilder builder = factory.newDocumentBuilder();
        // Запарсили XML, создав структуру Document. Теперь у нас есть доступ ко всем элементам, каким нам нужно.
        Document document = builder.parse(new File(fileName));
        // Получение списка всех элементов tlkFile внутри корневого элемента (getDocumentElement возвращает ROOT элемент XML файла).
        NodeList tlkFileElements = document.getDocumentElement().getElementsByTagName("tlkFile");
        /* read and store TLK Tool version, which was used to create the XML file */
        String toolVersion = String.valueOf(document.getDocumentElement().getAttributes().getNamedItem("TLKToolVersion").getNodeValue());
        if (toolVersion != null) {
            inputFileVersion = toolVersion;
        }

        // Получение списка всех элементов string внутри корневого элемента.
        NodeList stringElements = document.getDocumentElement().getElementsByTagName("string");
        for (int i=0; i< stringElements.getLength(); i++) {
            int id = 0, position = 0;
            String data = "";
            NodeList childNodes = stringElements.item(i).getChildNodes();

            for (int j=0; j<childNodes.getLength(); j++) {
                if (childNodes.item(j).getNodeName().equals("id")) {
                    id = Integer.parseInt(childNodes.item(j).getTextContent());
                } else if (childNodes.item(j).getNodeName().equals("position")) {
                    position = Integer.parseInt(childNodes.item(j).getTextContent());
                } else if (childNodes.item(j).getNodeName().equals("data")) {
                    data = childNodes.item(j).getTextContent();
                }
            }
            data = data.replace("\r\n", "\n");
            /* every String should be NULL-terminated */
            if (id >= 0) {
                data += '\0';
            }
            /* only add debug info if we are in debug mode and StringID is positive AND it's localizable */
            inputData.add(id >= 0 && debugVersion && (id & 0x8000000) != 0x8000000 ?
                    new TlkEntry(id, position, "(#" + id + ") " + data) :
                    new TlkEntry(id, position, data)
            );
        }

        /* code for XML files created BEFORE v. 1.0.3 */
        String lastEntryFixVersion = "1.0.3";

        /* check if someone isn't loading the bugged version < 1.0.3 */
        if (compareVersionStrings(inputFileVersion, lastEntryFixVersion, "<")) {
            throw new RuntimeException();
        }
    }

    private static boolean compareVersionStrings(String ver1, String ver2, String sign) {
        String[] arr1 = ver1.split("\\.");
        String[] arr2 = ver2.split("\\.");
        int len = Math.max(arr1.length, arr2.length);
        int[] arr1New = new int[len];
        int[] arr2New = new int[len];

        for (int i=0; i<len; i++) {
            arr1New[i] = arr1.length > i ? Integer.parseInt(arr1[i]) : 0;
            arr2New[i] = arr2.length > i ? Integer.parseInt(arr2[i]) : 0;
        }

        if ("<".equals(sign)) {
            for (int i = 0; i < len; i++) {
                if (arr1New[i] > arr2New[i]) {
                    return false;
                } else if (arr1New[i] < arr2New[i]) {
                    return true;
                }
            }
            return false;
        } else if (">".equals(sign)) {
            for (int i = 0; i < len; i++) {
                if (arr1New[i] > arr2New[i]) {
                    return true;
                } else if (arr1New[i] < arr2New[i]) {
                    return false;
                }
            }
            return false;
        } else if ("=".equals(sign)) {
            for (int i = 0; i < len; i++) {
                if (arr1New[i] > arr2New[i]) {
                    return false;
                } else if (arr1New[i] < arr2New[i]) {
                    return false;
                }
            }
            return true;
        }
        throw new RuntimeException("Set valid sign( '<' '>' '=' )");
    }

    /* maybe will be finished in the future */
//    private void LoadTxtInputData(String fileName) throws FileNotFoundException, UnsupportedEncodingException {
//
//        BufferedReader streamReader = new BufferedReader(new InputStreamReader(new FileInputStream(fileName)));
//        String line;
//        int i = 1;
//
//        while (streamReader.Peek() != -1) {
//            line = streamReader.readLine();
//            System.out.println(i++);
//            System.out.println(line);
//            String[] words = line.split(":");
//            System.out.println(words.length+" words in text:");
//            for (String s : words) {
//                System.out.println(s + " | ");
//            }
//            System.out.println();
//        }
//        streamReader.close();
//    }

    /** <summary>
     *       Creates Huffman Tree based on data from memory.
     *       For every character in text data, a corresponding Huffman Code is prepared.
     *       Source: http://en.wikipedia.org/wiki/Huffman_coding
     *   </summary>
     */
    private void PrepareHuffmanCoding() {
        frequencyCount.clear();
        for (TlkEntry entry : inputData) {
            if (entry.stringID < 0) {
                continue;
            }
            for (char c : entry.data.toCharArray()) {
                if (!frequencyCount.containsKey(c)) {
                    frequencyCount.put(c, 0);
                }
                frequencyCount.put(c, frequencyCount.get(c) + 1);
            }
        }
        frequencyCount.forEach((key, value) -> huffmanTree.add(new HuffmanNode(key, value)));
        BuildHuffmanTree();
        BuildCodingArray();
        // DebugTools.LoadHuffmanTree(_huffmanCodes);
        // DebugTools.PrintLookupTable();
    }

    /** <summary>
     *       Standard implementation of builidng a Huffman Tree
     * </summary>
     */
    private void BuildHuffmanTree() {
        while (huffmanTree.size() > 1) {
            /* sort Huffman Nodes by frequency */
            huffmanTree.sort(HuffmanCompression::CompareNodes);
            HuffmanNode parent = new HuffmanNode(huffmanTree.get(0), huffmanTree.get(1));
            huffmanTree.remove(0);
            huffmanTree.remove(0);
            huffmanTree.add(parent);
        }
    }

    /** <summary>
     *       Using Huffman Tree (created with BuildHuffmanTree method), generates a binary code for every character.
     *   </summary>
     */
    private void BuildCodingArray() {
        /* stores a binary code */
        List<Boolean> currentCode = new LinkedList<>();
        HuffmanNode currenNode = huffmanTree.get(0);
        TraverseHuffmanTree(currenNode, currentCode);
    }

    /** <summary>
     *       Recursively traverses Huffman Tree and generates codes
     *   </summary>
     *   <param name="node"></param>
     *   <param name="code"></param>
     */
    private void TraverseHuffmanTree(HuffmanNode node, List<Boolean> code) {
        /* check if both sons are null */
        if (node.left == node.right) {
            boolean[] arr = new boolean[code.size()];
            for (int i = 0; i < code.size(); i++) {
                arr[i] = code.get(i);
            }
            BitArray ba = new BitArray(arr);
            huffmanCodes.put(node.data, ba);
        } else {
            /* adds 0 to the code - process left son*/
            code.add(false);
            TraverseHuffmanTree(node.left, code);
            code.remove(code.size() - 1);

            /* adds 1 to the code - process right son*/
            code.add(true);
            TraverseHuffmanTree(node.right, code);
            code.remove(code.size() - 1);
        }
    }

    /** <summary>
     *       Converts a Huffman Tree to it's binary representation used by TLK format of Mass Effect 2.
     *   </summary>
     *   <returns></returns>
     */
    private List<Integer> ConvertHuffmanTreeToBuffer() {
        Queue<HuffmanNode> q = new ArrayDeque<>();
        Map<Integer, HuffmanNode> indices = new HashMap<>();

        int index = 0;
        q.add(huffmanTree.get(0));

        while (q.size() > 0) {
            HuffmanNode node = q.remove();
            /* if it's a leaf - set it's ID to reflect char data the node contains */
            if (node.left == node.right) {
                /* store the char data */
                node.id = -1 - node.data;

                /* that's how it's going to be decoded when parsing TLK file:
                 * char c = ru.BitConverter.ToChar(ru.BitConverter.GetBytes(0xffff - node.ID), 0); */
            } else {
                node.id = index++;
                indices.put(node.id, node);
            }
            if (node.right != null)
                q.add(node.right);
            if (node.left != null)
                q.add(node.left);
        }

        List<Integer> output = new LinkedList<>();

        indices.values().forEach(node -> {
            output.add(node.left.id);
            output.add(node.right.id);
        });
        return output;
    }

    /** <summary>
     *       Converts bits in a main.java.ru.BitArray to an array with bytes.
     *       Such array is ready to be written to a file.
     *   </summary>
     *   <param name="bitsList"></param>
     *   <param name="bitsCount"></param>
     *   <returns></returns>
     */
    private static byte[] BitArrayListToByteArray(List<BitArray> bitsList, int bitsCount) {
        final int BITSPERBYTE = 8;

        int bytesize = bitsCount / BITSPERBYTE;
        if (bitsCount % BITSPERBYTE > 0)
            bytesize++;

        byte[] bytes = new byte[bytesize];
        int bytepos = 0;
        int bitsRead = 0;
        byte value = 0;
        byte significance = 1;

        for (BitArray bits : bitsList) {
            int bitpos = 0;

            while (bitpos < bits.length()) {
                if (bits.get(bitpos)) {
                    value += significance;
                }
                ++bitpos;
                ++bitsRead;
                if (bitsRead % BITSPERBYTE == 0) {
                    bytes[bytepos] = value;
                    ++bytepos;
                    value = 0;
                    significance = 1;
                    bitsRead = 0;
                }
                else {
                    significance <<= 1;
                }
            }
        }
        if (bitsRead % BITSPERBYTE != 0) {
            bytes[bytepos] = value;
        }
        return bytes;
    }

    /** <summary>
     *      For sorting Huffman Nodes
     *  </summary>
     *  <param name="l1"></param>
     *  <param name="l2"></param>
     *  <returns></returns>
     */
    private static int CompareNodes(HuffmanNode l1, HuffmanNode l2) {
        return Integer.compare(l1.frequencyCount, l2.frequencyCount);
    }
}

File: python_ver/huffman_compression.py
import os
from xml.etree import ElementTree as ET

from collections import deque
from typing import Any

from helpers import bit_array_list_to_byte_array, compare_version_strings
from huffman_node import HuffmanNode


class HuffmanCompression:
    def __init__(self):
        self.input_file_version = '1.0.0.0'
        self.input_data = []
        self.frequency_count = {}
        self.huffman_tree = []
        self.huffman_codes = {}

    # Converts a Huffman Tree to it's binary representation used by TLK format of Mass Effect 2.
    def convert_huffman_tree_to_buffer(self) -> list[Any]:
        q = deque()
        indices = {}

        index = 0
        q.append(self.huffman_tree[0])

        while len(q) > 0:
            node = q.popleft()
            # if it's a leaf - set it's ID to reflect char data the node contains
            if node.left == node.right:
                # store the char data
                node.id = -1 - node.letter
                # that's how it's going to be decoded when parsing TLK file:
                # char c = ru.BitConverter.ToChar(ru.BitConverter.GetBytes(0xffff - node.ID), 0);
            else:
                node.id = index
                index += 1
                indices[node.id] = node
            if node.right is not None:
                q.append(node.right)
            if node.left is not None:
                q.append(node.left)
        output = []
        for node in indices.values():
            output.append(node.left.id)
            output.append(node.right.id)
        return output

    # Dumps data from memory to TLK compressed file format.
    # Compressed data should be read into memory first, by LoadInputData method.
    def save_to_tlk_file(self, dest_file: str) -> None:
        if os.path.isfile(dest_file):
            os.remove(dest_file)
        # converts Huffmann Tree to binary form
        tree_buffer = self.convert_huffman_tree_to_buffer()

        # preparing data and entries for writing to file;
        # entries list consists of pairs <string_id, offset>
        binary_data = []
        entries1 = {}
        entries2 = {}
        offset = 0

        for entry in self.input_data:
            str_id = entry.string_id
            if str_id < 0:
                if str_id not in entries1:
                    entries1[str_id] = int(entry.data)
                else:
                    entries2[str_id] = int(entry.data)
                continue

            if str_id not in entries1:
                entries1[str_id] = offset
            else:
                entries2[str_id] = offset

            # for every character in a string, put it's binary code into data array
            for ch in list(entry.data):
                binary_data.append(self.huffman_codes[ch])
                offset += len(self.huffman_codes[ch])

        # preparing TLK Header
        magic = 7040084
        ver = 3
        min_ver = 2
        entry_1_count = len(entries1)
        entry_2_count = len(entries2)
        tree_node_count = len(tree_buffer) // 2
        data_length = offset // 8
        if offset % 8 > 0:
            data_length += 1

        # writing TLK Header
        with open(dest_file, "w+") as f:
            f.write(str(magic))
            f.write(str(ver))
            f.write(str(min_ver))
            f.write(str(entry_1_count))
            f.write(str(entry_2_count))
            f.write(str(tree_node_count))
            f.write(str(data_length))

            # writing entries
            for k, v in entries1.items():
                f.write(k)
                f.write(v)
            for k, v in entries2.items():
                f.write(k)
                f.write(v)

            # writing HuffmanTree
            for el in tree_buffer:
                f.write(el)

            # writing data
            data = bit_array_list_to_byte_array(binary_data, offset)
            f.write(data)

    # Loads a file into memory and prepares for compressing it to TLK
    def load_input_data(self, file_name: str, ff: str):
        self.input_data = []
        self.load_xml_input_data(file_name)
        # sorted(self.input_data, key=lambda tlk_entry: tlk_entry.position)
        self.input_data.sort(key=lambda x: x.position)
        self.prepare_huffman_coding()

    def prepare_huffman_coding(self):
        self.frequency_count = {}
        for entry in self.input_data:
            if entry.string_id < 0:
                continue
            if entry.data is not None:
                for ch in list(entry.data):
                    if ch not in self.frequency_count.keys():
                        self.frequency_count[ch] = 0
                    self.frequency_count[ch] = self.frequency_count[ch] + 1

        # here we have frequency of each char at TLK file
        for k, v in self.frequency_count.items():
            self.huffman_tree.append(HuffmanNode(d=k, freq=v))
        self.build_huffman_tree()
        self.build_coding_array()

    # Using Huffman Tree (created with build_huffman_tree method), generates a binary code for every character.
    def build_coding_array(self):
        # stores a binary code
        currend_code = []
        current_node = self.huffman_tree[0]
        self.traverse_huffman_tree(current_node, currend_code)

    # Recursively traverses Huffman Tree and generates codes
    def traverse_huffman_tree(self, node, code):
        # check if both sons are None
        if node.left == node.right:
            # arr = []
            # for i in range(len(code)):
            #     arr.append(code[i])
            # ba = bit_array.BitArray(bits=arr)
            self.huffman_codes[node.letter] = ''.join(code)
        else:
            # adds 0 to the code - process left son
            code.append('0')
            self.traverse_huffman_tree(node.left, code)
            del code[len(code) - 1]

            # adds 1 to the code - process right son
            code.append('1')
            self.traverse_huffman_tree(node.right, code)
            del code[len(code) - 1]

    def build_huffman_tree(self):
        while len(self.huffman_tree) > 1:
            # sort Huffman Nodes by frequency
            self.huffman_tree.sort(key=lambda x: x.frequency_count)

            parent = HuffmanNode(left=self.huffman_tree[0], right=self.huffman_tree[1])
            del self.huffman_tree[0]
            del self.huffman_tree[0]
            self.huffman_tree.append(parent)

    # Loads data from XML file into memory
    def load_xml_input_data(self, file_name: str):
        tree = ET.parse(file_name)
        root = tree.getroot()

        # read and store TLK Tool version, which was used to create the XML file
        self.input_file_version = root.attrib['TLKToolVersion']

        # get list of all <string> elements inside <tlkFile> tag
        for child in root:
            id = 0
            position = 0
            data = ''
            for child_node in child:
                if child_node.tag == 'id':
                    id = int(child_node.text)
                elif child_node.tag == 'position':
                    position = int(child_node.text)
                elif child_node.tag == 'data':
                    data = child_node.text
            if data is not None:
                data.replace('\r\n', '\n')

                # every string should be NONE-terminated
                if id >= 0:
                    data += '\0'

            # only add debug info if we are in debug mode and StringID is positive AND it's localizable
            tlk_e = TlkEntry(id, position, data)
            self.input_data.append(tlk_e)
            # if id >= 0 && debug_version && (id & 0x8000000) != 0x8000000:
            #     tlk_e = tlk_entry(id, position, "(#" + id + ") " + data)
            #     self.input_data.append(tlk_e)
            # else:
            #     tlk_e = tlk_entry(id, position, data)
            #     self.input_data.append(tlk_e)

        # code for XML files created BEFORE v. 1.0.3
        last_entry_fix_version = '1.0.3'

        # check if someone isn't loading the bugged version < 1.0.3
        if compare_version_strings(self.input_file_version, last_entry_fix_version, '<'):
            raise Exception('version < 1.0.3')


class TlkEntry:
    def __init__(self, id: int, pos: int, data: str):
        self.string_id = id
        self.position = pos
        self.data = data
